### PR TITLE
Ib/53 set publish err reply

### LINKED ISSUE #53: The publish response object (ErrReply) should be used to communicate errors
Today errors that occur are not sent back to the framework in the response object.

```
err := p.plugin.Publish(metrics, cfg)
if err != nil {
	return &rpc.ErrReply{}, err
}
```

### PR DESCRIPTION
**Pull request by jcooklin**
Fixes #53 

Summary: 
- [x] populates ErrReply on error
- [x] Adds gitignore, makefile and supporting build and test scripts
 - make supports targets for testing, building examples, cleaning
 - plugin folders renamed to adhere to convention and make using common build and test scripts possible with only minimal changes
- [x] Updated the file publisher example to return an error if provided the bool config 'return_error' 
 
Testing:
- [x] manual
- [x] unit

Related: https://github.com/intelsdi-x/snap/issues/1450

### CODE REVIEW

File: scripts/build_examples.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+__proj_dir="$(dirname "$__dir")"
+
+# shellcheck source=scripts/common.sh
+. "${__dir}/common.sh"
+
+_info "project path: ${__proj_dir}"
+
+git_version=$(_git_version)
+go_build=(go build -ldflags "-w -X main.gitversion=${git_version}")
```
Comment (IRCody) [inline scripts/build_examples.sh]: Some of this stuff (this in particular) seems duplicated from build_example? am I missing where this is used here?

Comment (jcooklin) [inline scripts/build_examples.sh]: Good catch.  I removed it from `build_examples.sh` and updated `build_example.sh`.

File: v1/plugin/publisher_proxy.go [hunk 2]
```diff
@@ -42,8 +42,8 @@ func (p *publisherProxy) Publish(ctx context.Context, arg *rpc.PubProcArg) (*rpc
 	cfg := fromProtoConfig(arg.Config)
 	err := p.plugin.Publish(metrics, cfg)
 	if err != nil {
-		return &rpc.ErrReply{}, err
-	} else {
-		return &rpc.ErrReply{}, nil
+		return &rpc.ErrReply{Error: err.Error()}, nil
 	}
+	return &rpc.ErrReply{}, nil
```
Comment (kindermoumoute) [inline v1/plugin/publisher_proxy.go]: Does this method need to return a 2nd argument anymore?

File: examples/snap-plugin-publisher-file/file/file.go [hunk 3]
```diff
@@ -49,6 +50,9 @@ func (f FPublisher) Publish(mts []plugin.Metric, cfg plugin.Config) error {
 	if err != nil {
 		return err
 	}
+	if val, err := cfg.GetBool("return_error"); err == nil && val {
+		return errors.New("Houston we have a problem")
```
Comment (kindermoumoute) [inline examples/snap-plugin-publisher-file/file/file.go]: `<joke>` Not consistent with "Houston, we have a problem!" in the rand collector `</joke>`

**Comment by candysmurf** [comment]
@jcooklin, would you please resolve the conflict once you have a chance?

**Comment by jcooklin** [comment]
Rebased.  cc: @IRCody @candysmurf

**Comment by kjlyon** [comment]
@jcooklin, What is status of this PR? Is this ready to be merged?

**Pull request merged**